tests/src/serverTestArrayPersistence.cpp

```diff
@@ -0,0 +1,140 @@
+};
+
+ReferenceTestApplication GlobalFixture::referenceTestApplication{BOOST_STRINGIZE(BOOST_TEST_MODULE), alen};
+std::string GlobalFixture::rpcNo;
+std::string GlobalFixture::bpn;
+
+CTK_BOOST_GLOBAL_FIXTURE(GlobalFixture)
+
+// the array must have testStartValue+i at index i.
+template<class T>
+static bool testArrayContent1(std::vector<T>& array, std::string const& propertyName, T testStartValue, T delta) {
+  bool isOK = true;
+  T currentTestValue = testStartValue;
+  size_t index = 0;
+  for(auto val : array) {
+    if(std::fabs(val - currentTestValue) > 0.001) {
+      if(isOK) {
+        std::cout << "Array " << propertyName << " does not contain expected values. First mismatching data at index "
+                  << index << ": " << val << " != " << currentTestValue << std::endl;
+      }
+      isOK = false;
+    }
+    currentTestValue += delta;
+    ++index;
+  }
+  return isOK;
+}
+template<class T>
+static bool testArrayContent(std::string const& propertyName, T testStartValue, T delta) {
+  auto array = DoocsServerTestHelper::doocsGetArray<T>(propertyName);
+  return testArrayContent1(array, propertyName, testStartValue, delta);
+}
+
+/// Using prepared persistence file, start server, check restored values, change values, trigger persisting, compare file with expectation
+BOOST_AUTO_TEST_CASE(testArrayWrite) {
+  GlobalFixture::referenceTestApplication.runMainLoopOnce();
+
+  std::string const& propertyAddress = "//INT/MY_RENAMED_INTARRAY";
+  checkDataType(propertyAddress, DATA_A_INT);
+
+  CHECK_WITH_TIMEOUT(testArrayContent<float>("//INT/MY_RENAMED_INTARRAY", 100, 1) == true);
+  CHECK_WITH_TIMEOUT(testArrayContent<float>("//DOUBLE/FROM_DEVICE_ARRAY", 200.1, 1) == true);
+
+  std::vector<int> vint(alen);
+  std::vector<double> vdouble(alen);
+  for(int i = 0; i < alen; i++) {
+    vint[i] = 150 + i;
+    vdouble[i] = 250.3 + i;
+  }
+  DoocsServerTestHelper::doocsSet("//INT/TO_DEVICE_ARRAY", vint);
+  DoocsServerTestHelper::doocsSet("//DOUBLE/TO_DEVICE_ARRAY", vdouble);
+
+  // run the application loop.
+  GlobalFixture::referenceTestApplication.runMainLoopOnce();
+
+  // check changed values
+  usleep(100000);
+  CHECK_WITH_TIMEOUT(testArrayContent<float>("//INT/MY_RENAMED_INTARRAY", 150, 1) == true);
+  CHECK_WITH_TIMEOUT(testArrayContent<float>("//DOUBLE/FROM_DEVICE_ARRAY", 250.3, 1) == true);
+
+  // trigger array storing
+  DoocsServerTestHelper::doocsSet<bool>("//ARRAY_PERSISTENCE_TEST._SVR/SVR.SAVE", true);
+  // wait till disk file written
+  usleep(500000);
+
+  // check persisted file against known values
+  std::vector<int> vint2(alen);
+  std::vector<double> vdouble2(alen);
+  vectorFromFile(vint2, "hist/INT-TO_DEVICE_ARRAY.intarray");
+  vectorFromFile(vdouble2, "hist/DOUBLE-TO_DEVICE_ARRAY.doublearray");
+  CHECK_WITH_TIMEOUT(testArrayContent1<int>(vint2, "//INT/MY_RENAMED_INTARRAY", 150, 1) == true);
```

Comment:
`CHECK_WITH_TIMEOUT` is pointless in this place, since you are comparing an array which has been read from file previously. To fulfill the previous comment, you could write it like this:

```
CHECK_WITH_TIMEOUT(vectorFromFile(vint2, "hist/INT-TO_DEVICE_ARRAY.intarray"), testArrayContent1<int>(vint2, "//INT/MY_RENAMED_INTARRAY", 150, 1) == true);
```

Besides: I find the mentioning of the property name here confusing, since it does not check the actual property. Maybe change the string into e.g. `"//INT/MY_RENAMED_INTARRAY from file"`. This will both improve the error message if it fails and improves readability of the code (I had to check `testArrayContent1()` carefully before understanding it is not reading from that property). Also please find a better name for `testArrayContent()`/`testArrayContent1()`, e.g. `testArrayContentFromProperty()` and `testArrayContent()` - note I swapped the meaning of `testArrayContent()` here...

Reply:
fixed.
I implemented an explicit loop for the timeout behavior since CHECK_WITH_TIMEOUT would try 10000 times, which I believe would take too long.

Reply:
If it takes long in an error case (even a few minutes) this is annoying but OK. It is important that the test is fast in the success case, so the time per iteration counts.